c_binding/lsm_local_disk.c

```diff
@@ -1042,3 +1128,170 @@ int LSM_DLL_EXPORT lsm_local_disk_led_status_get(const char *disk_path,
     }
     return rc;
 }
+
+int lsm_local_disk_link_speed_get(const char *disk_path,
+                                  const char **link_speed, lsm_error **lsm_err)
+{
+    int rc = LSM_ERR_OK;
+    int tmp_rc = LSM_ERR_OK;
+    lsm_error *tmp_lsm_err = NULL;
+    lsm_disk_link_type link_type = LSM_DISK_LINK_TYPE_UNKNOWN;
+    int fd = -1;
+    char err_msg[_LSM_ERR_MSG_LEN];
+    uint8_t vpd_data[_SG_T10_SPC_VPD_MAX_LEN];
+    struct _sg_t10_vpd_ata_info *ata_info = NULL;
+    int sata_link_speed = 0;
+    int sas_link_speed = 0;
+    uint8_t sas_mode_sense[_SG_T10_SPC_MODE_SENSE_MAX_LEN];
+    char sas_addr[_SG_T10_SPL_SAS_ADDR_LEN];
+    unsigned int host_no = UINT_MAX;
+    unsigned int fc_speed = _FC_SPEED_UNKNOWN;
+    size_t fc_speed_str_len = 0;
+    char *tmp_str = NULL;
+
+    _lsm_err_msg_clear(err_msg);
+    rc = _check_null_ptr(err_msg, 3 /* argument count */, disk_path,
+                         link_speed, lsm_err);
+
+    if (rc != LSM_ERR_OK) {
+        /* set output pointers to NULL if possible when facing error in case
+         * application use output memory.
+         */
+        if (link_speed != NULL)
+            *link_speed = NULL;
```

Comment:
@Mashimiao No it should be fine. The client shouldn't have allocated memory for us to fill. Take a look at the unit test for a usage sample.